File: src/ReetCode/Easy/1to100/Q67/Q67_Add Binary.py
"""
# Author    : Lee In Bok 
# Date      : 2023.11.04(Sat)
# Spend Time: 47m 50s
# Runtime   : 37 ms (Beats 81.76%)
# Memory    : 16.2 MB (Beats 81.11%)
# Algoritm  : String
"""

# class Solution:
#     def addBinary(self, a: str, b: str) -> str:
#         return bin(int(a, 2) + int(b, 2))[2:]

class Solution:
    def addBinary(self, a: str, b: str) -> str:
        carry, result = False, []
        x, y = len(a) - 1, len(b) - 1

        if a == b == "0":
            return "0"

        while 0 <= x and 0 <= y:
            if a[x] == b[y] == "1":
                result.append("1" if carry else "0")
                carry = True
            elif a[x] == b[y] == "0":
                result.append("1" if carry else "0")
                carry = False
            else:
                result.append("0" if carry else "1")
                carry = True if carry else False

            x -= 1
            y -= 1
        
        carry = self.checkRest(a, x, carry, result)
        carry = self.checkRest(b, y, carry, result)

        if result[-1] == "0" or carry:
            result.append("1")

        return ''.join(reversed(result))
    
    def checkRest(self, arr, idx, carry, result):
        while 0 <= idx:
            if arr[idx] == "1":
                result.append("0" if carry else "1")
                carry = True if carry else False
            else:
                result.append("1" if carry else "0")
                carry = False

            idx -= 1

        return carry

s = Solution()
result = s.addBinary("101111", 
                     "10")

print(result)

File: src/ReetCode/Easy/1to100/Q67/Q67_AddBinary.java
/**
* Author    : Lee In Bok 
* Date      : 2023.11.04(Sat)
* Spend Time: 47m 50s
* Runtime   : 2 ms (Beats 55.51%)
* Memory    : 41 MB (Beats 70.52%)
* Algoritm  : String
 */

import java.util.ArrayList;
import java.util.List;

public class Q67_AddBinary {
    public static void main(String[] args) {
        Solution s = new Solution();
        String result = s.addBinary("1010", "1011");

        System.out.println(result);
    }
}

// class Solution {
//     public String addBinary(String a, String b) {
//         return Integer.toBinaryString(Integer.parseInt(a, 2) + Integer.parseInt(b, 2));
//     }
// }

class Solution {

    public final String ZERO = "0";
    public final String ONE = "1";

    public String addBinary(String a, String b) {
        if(a.equals(ZERO) && b.equals(ZERO)){
            return a;
        }

        boolean carry = false;
        List<String> result = new ArrayList<>();
        int x = a.length() - 1;
        int y = b.length() - 1;

        while(0 <= x && 0 <= y){
            if(a.charAt(x) == '1' && b.charAt(y) == '1'){
                result.add(carry ? ONE : ZERO);
                carry = true;
            } else if(a.charAt(x) == '0' && b.charAt(y) == '0'){
                result.add(carry ? ONE : ZERO);
                carry = false;               
            } else {
                result.add(carry ? ZERO : ONE);
                carry = carry ? carry : false;
            }

            x--;
            y--;
        }

        carry = checkrest(a, x, carry, result);
        carry = checkrest(b, y, carry, result);

        if(carry || result.get(result.size() - 1).equals(ZERO)){
            result.add(ONE);
        }

        StringBuilder sb = new StringBuilder();

        for(String c : result){
            sb.append(c);
        }

        return sb.reverse().toString();
    }

    public boolean checkrest(String s, int idx, boolean carry, List<String> result){
        while(0 <= idx){
            if(s.charAt(idx) == '1'){
                result.add(carry ? ZERO : ONE);
                carry = carry ? carry : false;
            } else {
                result.add(carry ? ONE : ZERO);
                carry = false;
            }
            idx--;
        }

        return carry;
    }
}